adam/learner/relations.py

```diff
@@ -133,3 +155,218 @@ def _update_hypothesis(
+        # Furthermore we only need two nodes, because relations only involve two objects.
+        return (
+            len(
+                language_perception_semantic_alignment.perception_semantic_alignment.semantic_nodes
+            )
+            > 2
+        )
+
+    def _preprocess_scene(
+        self, perception_semantic_alignment: PerceptionSemanticAlignment
+    ) -> PerceptionSemanticAlignment:
+        return perception_semantic_alignment
+
+    def log_hypotheses(self, log_output_path: Path) -> None:
+        logging.info(
+            "Logging %s hypotheses to %s",
+            len(self._concept_to_hypotheses_and_scores),
+            log_output_path,
+        )
+        for (
+            concept,
+            hypotheses_to_scores,
+        ) in self._concept_to_hypotheses_and_scores.items():
+            for (i, (hypothesis, score)) in enumerate(hypotheses_to_scores.items()):
+                hypothesis.render_to_file(
+                    concept.debug_string,
+                    log_output_path / f"{concept.debug_string}.{i}.score_{score}",
+                )
+
+        logging.info(
+            "Logging %s lexicon items to %s", len(self._lexicon), log_output_path
+        )
+        for (
+            surface_template,
+            meaning_as_perception_graph_template,
+        ) in self._lexicon.items():
+            meaning_as_perception_graph_template.render_to_file(
+                surface_template.to_short_string(),
+                log_output_path / f"lexicon-{surface_template.to_short_string()}",
+            )
+
+
+class ImmutableSet(object):
+    pass
+
+
+def _extract_candidate_relations(
+    whole_scene_perception_graph: PerceptionGraph,
+    relation_object_1: ObjectSemanticNode,
+    relation_object_2: ObjectSemanticNode,
+) -> Sequence[PerceptionGraph]:
+    # The directions of edges in the perception graph are not necessarily meaningful
+    # from the point-of-view of hypothesis generation, so we need an undirected copy
+    # of the graph.
+    perception_digraph = whole_scene_perception_graph.copy_as_digraph()
+    perception_graph_undirected = perception_digraph.to_undirected(
+        # as_view=True loses determinism
+        as_view=False
+    )
+
+    output_graphs = []
+
+    # The core of our hypothesis for the semantics of a preposition is all nodes
+    # along the shortest path between the two objects involved in the perception graph.
+    for hypothesis_spine_nodes in all_shortest_paths(
+        perception_graph_undirected, relation_object_2, relation_object_1
+    ):
+        # Along the core of our hypothesis we also want to collect the predecessors and successors
+        hypothesis_nodes_mutable = []
+        for node in hypothesis_spine_nodes:
+            if node not in {relation_object_1, relation_object_2}:
+                for successor in perception_digraph.successors(node):
+                    if not isinstance(successor, ObjectPerception):
```

Comment:
Do we also want to exclude `ObjectSemanticNode` s along the path and explicitly include the relation objects?